File: clp/minimal_loc/java.pl
mapMember(K, V, map(Pairs, _, _)) :-
    member(pair(K, V), Pairs).
mapAdd(map(InPairs, R1, R2), K, V, map(OutPairs, R1, R2)) :-
    mapAddInternal(InPairs, R1, R2, K, V, OutPairs).
mapAddInternal([], _, _, K, V, [pair(K, V)]).
mapAddInternal([pair(K1, _)|Rest], KeysSame, _, K2, V, [pair(K2, V)|Rest]) :-
    call(KeysSame, K1, K2).
mapAddInternal([Pair|InRest], KeysSame, KeysDifferent, K1, V, [Pair|OutRest]) :-
    Pair = pair(K2, _),
    call(KeysDifferent, K1, K2),
    mapAddInternal(InRest, KeysSame, KeysDifferent, K1, V, OutRest).
makeEqualsMap(map([],
                  lambda([K1, K2], K1 == K2),
                  lambda([K1, K2], K1 \== K2))).
hierarchyGet(Spec, ClassName, DirectParents) :-
    once(lambda([], member(class_spec(ClassName, DirectParents, _, _), Spec))).
allParents(Spec, Of, Result) :-
    hierarchyGet(Spec, Of, DirectParents),
    member(DirectParent, DirectParents),
    (Result = DirectParent;
     allParents(Spec, DirectParent, Result)).
once(Call) :-
    call(Call),
    !.
parentChild(Spec, Parent, Child) :-
    hierarchyGet(Spec, Child, DirectParents),
    member(Parent, DirectParents),
    !.
parentChild(Spec, Parent, Child) :-
    hierarchyGet(Spec, Child, DirectParents),
    member(DirectParent, DirectParents),
    parentChild(Spec, Parent, DirectParent),
    !.
typesCompatibleParentChild(_, Type, Type) :- !.
typesCompatibleParentChild(Spec, class_type(ParentName), class_type(ChildName)) :-
    parentChild(Spec, ParentName, ChildName),
    !.
makeExpressionsOfTypes(_, _, _, [], []).
makeExpressionsOfTypes(Spec, Size, Env, [Type|Types], [Expression|Expressions]) :-
    makeExpression(Spec, Size, Env, Expression, ActualType),
    typesCompatibleParentChild(Spec, Type, ActualType),
    makeExpressionsOfTypes(Spec, Size, Env, Types, Expressions).
findCompatibleMethod(Spec, ClassName, MethodSpec) :-
    member(class_spec(ClassName, _, _, MethodSpecs), Spec),
    member(MethodSpec, MethodSpecs).
returnTypeOk(some(_), _) :- !.
returnTypeOk(_, yes).
makeCall(Spec, Size, Env, VoidOk, call(Base, MethodName, Params), ReturnType) :-
    makeExpression(Spec, Size, Env, Base, ExpressionType),
    ExpressionType = class_type(MyClassName), % intentionally mirroring Scala
    (MethodClassName = MyClassName;
     allParents(Spec, MyClassName, MethodClassName)),
    findCompatibleMethod(Spec, MethodClassName, method_spec(MethodName, ParamTypes, ReturnType)),
    returnTypeOk(ReturnType, VoidOk),
    makeExpressionsOfTypes(Spec, Size, Env, ParamTypes, Params).
decBound(In, Out) :-
    In > 0,
    Out is In - 1.
makeExpression(_, _, _, int_literal(0), int_type).
makeExpression(_, _, _, boolean_literal(no), boolean_type).
makeExpression(_, _, _, double_literal(0), double_type).
makeExpression(_, _, _, string_literal('foo'), class_type(class_name('String'))).
makeExpression(_, _, env(Map), variable_expression(Variable), Type) :-
    mapMember(Variable, Type, Map).
makeExpression(Spec, Size1, Env, new_expression(ClassName, Params), class_type(ClassName)) :-
    decBound(Size1, Size2),
    member(class_spec(ClassName, _, Constructors, _), Spec),
    member(constructor_spec(ParamTypes), Constructors),
    makeExpressionsOfTypes(Spec, Size2, Env, ParamTypes, Params).
makeExpression(Spec, Size1, Env, call_expression(Call), ReturnType) :-
    decBound(Size1, Size2),
    makeCall(Spec, Size2, Env, no, Call, RawReturnType),
    RawReturnType = some(ReturnType).
yolo_UNSAFE_atom_concat_number(Atom, Int, Result) :-
    number_atom(Int, IntAtom),
    atom_concat(Atom, IntAtom, Result).
makeStatement(Spec, Size, Env, StatementNum,
              variable_declaration(InitializerType, Variable, Initializer),
              env(NewMap)) :-
    makeExpression(Spec, Size, Env, Initializer, InitializerType),
    yolo_UNSAFE_atom_concat_number('x', StatementNum, VariableName),
    Variable = variable(VariableName),
    Env = env(Map),
    mapAdd(Map, Variable, InitializerType, NewMap).
makeStatement(Spec, Size, Env, _,
              call_statement(Call),
              Env) :-
    makeCall(Spec, Size, Env, yes, Call, _).
makeStatements(_, _, _, 0, []).
makeStatements(Spec, Size, Env, NumStatements, [Statement|Statements]) :-
    NumStatements > 0,
    makeStatement(Spec, Size, Env, NumStatements, Statement, NewEnv),
    NewNumStatements is NumStatements - 1,
    makeStatements(Spec, Size, NewEnv, NewNumStatements, Statements).
makeTest(Bound, java_test(Statements)) :-
    defaultSpec(Spec),
    makeEqualsMap(EmptyMap),
    makeStatements(Spec, Bound, env(EmptyMap), 1, Statements).
java_main(Size, Test) :-
    makeTest(Size, Test).

File: src/bench/test/scala/scife/enumeration/benchmarks/streams/java.scala
package scife.enumeration.benchmarks.streams

import scife.enumeration.benchmarks.java._

import scala.collection.immutable.Stream

// Generates Java expressions and statements with the intention
// of testing APIs.  Assumes there are existing classes around.
// Only interested in creating new objects and calling methods.
// Does not handle generics.

object StreamJavaGeneration {
  def disjuncts[A](streams: Stream[A]*): Stream[A] = {
    streams.reduceRight(_ ++ _)
  }

  def main(args: Array[String]) {
    if (args.length != 1) {
      println("Needs a size bound")
    } else {
      val bound = args(0).toInt
      val generator = apply()
      println("NUM ELEMENTS: " + generator.makeTest(bound).toList.size)
    }
  } // main

  def apply(): StreamJavaGeneration = {
    new StreamJavaGeneration(scife.enumeration.benchmarks.iterators.JavaGeneration.DEFAULT_SPEC)
  } // apply
}
import StreamJavaGeneration._

class StreamJavaGeneration(val spec: List[ClassSpec]) {
  import scife.enumeration.benchmarks.iterators.JavaGeneration.{
    Env,
    MIN_NUM_STATEMENTS,
    MAX_NUM_STATEMENTS
  }

  // for a given class, these are the parents
  val hierarchy: Map[ClassName, Set[ClassName]] =
    spec.map(s => (s.name, s.extendsImplements)).toMap

  def allParents(of: ClassName): Stream[ClassName] = {
    hierarchy.get(of) match {
      case Some(directParents) => {
	val asStream = directParents.toStream
	disjuncts(asStream,
		  asStream.flatMap(allParents))
      }
      case None => Stream.Empty
    }
  } // sameAndAllParents

  def parentChild(parent: ClassName, child: ClassName): Boolean = {
    hierarchy.get(child) match {
      case Some(directParents) => {
	if (directParents.contains(parent)) {
	  true
	} else {
	  directParents.exists(directParent => parentChild(parent, directParent))
	}
      }
      case None => false
    }
  } // parentChild

  def typesCompatibleParentChild(parent: Type, child: Type): Boolean = {
    (parent, child) match {
      case (`child`, `parent`) => true
      case (ClassType(parentName), ClassType(childName)) => parentChild(parentName, childName)
      case _ => false
    }
  } // typesCompatibleParentChild

  def makeExpressionsOfTypes(size: Int, env: Env, types: List[Type]): Stream[List[Expression]] = {
    types match {
      case Nil => Stream(List())
      case head :: tail => {
	for {
	  (expression, expressionType) <- makeExpression(size, env)
	  // given expression can be a subtype of what we expect
	  if typesCompatibleParentChild(head, expressionType)
	  rest <- makeExpressionsOfTypes(size, env, tail)
	} yield expression :: rest
      }
    }
  } // makeExpressionsOfTypes

  def makeCall(size: Int, env: Env, voidOk: Boolean): Stream[(Call, Option[Type])] = {
    for {
      (base, ClassType(myClassName)) <- makeExpression(size, env)
      methodClassName <- (myClassName #:: allParents(myClassName))
      MethodSpec(methodName, paramTypes, returnType) <- spec.find(_.name == methodClassName).get.methods.toStream
      if returnType.isDefined || voidOk
      params <- makeExpressionsOfTypes(size, env, paramTypes)
    } yield (Call(base, methodName, params), returnType)
  } // makeCall
  
  def makeExpression(size: Int, env: Env): Stream[(Expression, Type)] = {
    val baseCases: Stream[(Expression, Type)] =
      ((IntLiteral(0), IntType) #::
       (BooleanLiteral(false), BooleanType) #::
       (DoubleLiteral(0.0), DoubleType) #::
       (StringLiteral("foo"), ClassType(ClassName("String"))) #::
       env.toStream.map(pair => (VariableExpression(pair._1): Expression, pair._2)))

    if (size <= 0) {
      baseCases
    } else {
      disjuncts(baseCases,
		// new
		for {
		  ClassSpec(className, _, constructors, _) <- spec.toStream
		  ConstructorSpec(paramTypes) <- constructors.toStream
		  params <- makeExpressionsOfTypes(size - 1, env, paramTypes)
		} yield (NewExpression(className, params), ClassType(className)),
		// call - non-void
		for {
		  (call, Some(returnType)) <- makeCall(size - 1, env, false)
		} yield (CallExpression(call), returnType))
    }
  } // makeExpression
				 
  def makeStatement(size: Int, env: Env, statementNum: Int): Stream[(Statement, Env)] = {
    disjuncts(// variable declaration
              for {
		(initializer, initializerType) <- makeExpression(size, env)
	      } yield {
		val variable = Variable("x" + statementNum)
		(VariableDeclaration(initializerType, variable, initializer), env + (variable -> initializerType))
	      },
              // call - possibly void
              for {
		(call, _) <- makeCall(size, env, true)
	      } yield (CallStatement(call), env))
  } // makeStatement

  def makeStatements(size: Int, env: Env, numStatements: Int): Stream[List[Statement]] = {
    if (numStatements < 0) {
      Stream.Empty
    } else if (numStatements == 0) {
      Stream(List())
    } else {
      assert(numStatements > 0)
      for {
	(statement, newEnv) <- makeStatement(size, env, numStatements)
	rest <- makeStatements(size, newEnv, numStatements - 1)
      } yield statement :: rest
    }
  } // makeStatements

  def makeTest(size: Int): Stream[JavaTest] = {
    for {
      numStatements <- MIN_NUM_STATEMENTS.to(MAX_NUM_STATEMENTS).toStream
      statements <- makeStatements(size, Map(), numStatements)
    } yield JavaTest(statements.toSeq)
  } // makeTest
} // JavaGeneration

